vhost-device-scmi/src/devices/iio.rs

```diff
@@ -286,6 +296,97 @@ impl SensorT for IIOSensor {
         }
         Ok(result)
     }
+
+    fn get_notify_fd(&self) -> Option<RawFd> {
+        self.sensor.notify_dev.as_ref().map(|fd| fd.as_raw_fd())
+    }
+
+    fn reading_device(&self, device_index: u32) -> DeviceResult {
+        let mut result = vec![];
+
+        //Here we should create buffer which length is corresponding to iio:device;
+        //future change to self.get_data_format
+        //cat /sys/bus/iio/devices/iio:deviceX/scan_elements/in_XXX_type.
+        //for example le:s16/16>>0 , [x,y,z,(t)] each is signed 16, and Little Endian
+        //so one sample length is (3 or 4)*(16/8) byte
+        let sample_buffer_len = 8;
+        let mut buffer = vec![0u8; sample_buffer_len];
+        let mut file = self.sensor().notify_dev.as_ref().unwrap();
+
+        match file.read(&mut buffer) {
+            Ok(len) => {
+                if len > 0 {
+                    result.push(MessageValue::Unsigned(0)); //Agentid
+                    result.push(MessageValue::Unsigned(device_index)); //Sensorid
+                    let _time_stamp = i16::from_le_bytes(
+                        buffer[self.axes.len() * 2..self.axes.len() * 2 + 2]
+                            .try_into()
+                            .unwrap(),
+                    );
+                    for i in 0..self.axes.len() {
+                        //le:s16/16>>0
+                        let value =
+                            i16::from_le_bytes(buffer[i * 2..i * 2 + 2].try_into().unwrap());
+                        let value_i64 = value as i64;
+                        let sensor_value_low = (value_i64 & 0xffff_ffff) as i32;
+                        let sensor_value_high = (value_i64 >> 32) as i32;
+                        result.push(MessageValue::Signed(sensor_value_low));
+                        result.push(MessageValue::Signed(sensor_value_high));
+                        result.push(MessageValue::Unsigned(0));
+                        result.push(MessageValue::Unsigned(0));
+                    }
```

Comment:
Shouldn't the values be output basically the same way as in reading_get? Would it be possible to utilize reading_get here?

Reply:
It is quite different from function _reading_get_.
In function _reading_get_, it get data from _/sys/bus/iio/devices/iio:deviceX/in_XXX_raw_

But in notificaiton flow, packaged data is from "/dev/iio:deviceX" char device.
And this device support epoll to wait data, sysfs not.

Reply:
So the Linux IIO interfaces are quite different for explicit readings and for notifications. Readings use a separate file per each axis and notifications use a single file for all. I wonder about the other differences:
- Is there no scaling to be applied when reading the notification value (in the sense of iio:device/scale file)?
- Why do you read the value as i16 and then convert it to i64? Could the values be simply handled as i16 here and in reading_get? Or is it better to expect i64, to be future-proof?
- reading_get outputs u32, while reading_update i32. Looking into the SCMI specification, your version is correct. Should reading_get be fixed?

Reply:
> Is there no scaling to be applied when reading the notification value (in the sense of iio:device/scale file)?
According to scmi spec, frontend gets backend's iio axis configuration via cmd "**SENSOR_ AXIS_DESCRIPTION_GET**"
From spec **_4.7.2.6.1 Sensor Axis Descriptor_**, The bit[8] of **_uint32 axis_attributes_low_** represents that whether this sensor reports extended attributes of axis. Then extend value of "**_uint32 axis_resolution_**" will report resolution.

But function **_"fn axis_description(&self, axis: u32) -> Vec<MessageValue>"_** in currently common.rs, it always return zero. So it doesn't support getting scale now.

And what's need to note is that data from fs has same resolution with the data from /dev/iio:deviceX.
Function **reading_update** is just simply transmit data from dev without multiplied resolution. It is same with function **reading_get**.

Reply:
> reading_get outputs u32, while reading_update i32. Looking into the SCMI specification, your version is correct. Should reading_get be fixed?

Yes, you are right, according to spec **4.7.2.12.1 Sensor Reading Descriptor**,  the **sensor_value_low** and **sensor_value_high** should be both i32. This function should be fixed.

Reply:
> And what's need to note is that data from fs has same resolution with the data from /dev/iio:deviceX. Function **reading_update** is just simply transmit data from dev without multiplied resolution. It is different with function **reading_get**.

This is what I don't understand: Why does `reading_get` apply scaling while `reading_update` not?

Reply:
Is it perhaps because the scan_type format specification contains the scaling (shift), which is 0 in this case?

Reply:
> I don't know whether i64 is better, and why read_axis return i64.

I think I used i64 because this is what the SCMI specification supports. I think it makes sense in reading_get. As for reading_update, it could be simplified a bit assuming the supported value is i16 or left as it is. Up to you, depending what you decide to do about the supported formats.

Reply:
> according to spec **4.7.2.12.1 Sensor Reading Descriptor**, the **sensor_value_low** and **sensor_value_high** should be both i32. This function should be fixed.

Would you prefer to fix it in your patches or should I post the fix separately?

Reply:
> This is what I don't understand: Why does reading_get apply scaling while reading_update not?

funcation _reading_get_ is not implemented by us, in this function, it calls _read_axis_ to get data.
```
fn read_axis(&self, axis: &Axis) -> Result<i64, ScmiDeviceError> {
        let path_result = axis.path.clone().into_string();
        let mut value: i64 =
            read_number_from_file(Path::new(&(path_result.unwrap() + "_raw")))?.unwrap();
        let offset: Option<i64> = self.read_axis_offset(&axis.path)?;
        if let Some(offset_value) = offset {
            match value.checked_add(offset_value) {
                Some(new_value) => value = new_value,
                None => {
                    error!(
                        "IIO offset overflow in {:?}: {} + {}",
                        &axis.path,
                        value,
                        offset.unwrap()
                    );
                    return Err(ScmiDeviceError::GenericError);
                }
            }
        }
        let scale: Option<f64> = self.read_axis_scale(&axis.path)?;
        if let Some(scale_value) = scale {
            let exponent_scale = 10.0_f64.powi(i32::from(axis.custom_exponent));
            value = (value as f64 * (scale_value / exponent_scale)).round() as i64;
        }
        Ok(value)
    }
```
But I don't think the operation multiplied with scale_value should be done here. It should just return raw data from fs.
And the frontend's scale value should return by command **SENSOR_ DESCRIPTION_GET** or **SENSOR_ AXIS_DESCRIPTION_GET**

Then frontend can get real value by multiplied rawdata with scale 

It's my opinion.

Reply:
> Would you prefer to fix it in your patches or should I post the fix separately?
Both ok, I can update it with a single patch and rebase it in this pull request.

Reply:
> I can update it with a single patch and rebase it in this pull request.

OK, thank you.

Reply:
> But I don't think the operation multiplied with scale_value should be done here. It should just return raw data from fs.
> And the frontend's scale value should return by command **SENSOR_ DESCRIPTION_GET** or **SENSOR_ AXIS_DESCRIPTION_GET**

I think this is what `custom_exponent` function ensures. But IIRC there was no assurance that the IIO scale provided by the fs is a power of 10. Hence the additional scaling in reading_get/read_axis. Maybe it's only a theoretical problem, I don't know, but I wanted to be on the safe side. Whatever we decide to do, I'd like to have reading_get and reading_update doing the same thing, having divergent computations would be confusing and maybe incorrect in one of them.

Reply:
I find a problem that, in our embedded environment, we use ASM330LHH sensor, and it does not provide
suhc "**in_anglvel_X_scale**" for user, it only provide "**in_anglvel_scale**".
It will cause that function **_"read_axis_scale"_** always return _**None**_. It finally causes that **_scale_** is _**None**_

Does function "**_reading_get_**" should consider this?

Reply:
I make a modification like this to deal this situation.

```
diff --git a/vhost-device-scmi/src/devices/iio.rs b/vhost-device-scmi/src/devices/iio.rs
index 25e66ea..9066307 100644
--- a/vhost-device-scmi/src/devices/iio.rs
+++ b/vhost-device-scmi/src/devices/iio.rs
@@ -522,6 +522,7 @@ impl IIOSensor {
    fn read_axis_file<T: FromStr>(
        &self,
        path: &OsStr,
        name: &str,
    ) -> Result<Option<T>, ScmiDeviceError> {
         for value_path in [
             Path::new(&(String::from(path.to_str().unwrap()) + "_" + name)),
             &Path::new(&path).parent().unwrap().join(name),
+            Path::new(&format!("{}_{}",&String::from(path.to_str().unwrap())[..path.len()-2],name)),
         ]
         .iter()
         {
```
So that it can find all kinds of "scale" attribute

- /sys/bus/iio/devices/iio:device0/in_anglvel_y_scale
- /sys/bus/iio/devices/iio:device0/scale
- /sys/bus/iio/devices/iio:device0/in_anglvel_scale

Previous, it can only deal the first two cases.

Reply:
> I think this is what custom_exponent function ensures. But IIRC there was no assurance that the IIO scale provided by the fs is a power of 10. Hence the additional scaling in reading_get/read_axis. Maybe it's only a theoretical problem, I don't know, but I wanted to be on the safe side. Whatever we decide to do, I'd like to have reading_get and reading_update doing the same thing, having divergent computations would be confusing and maybe incorrect in one of them.

I analysis the scmi kernel and iio drivers, and your work is right.
Because the in _**axis_attributes_high**_, only has Exponent this single value. It means that in frontend scmi iio sysfs "in_XXX_scale" is must a value power-of-10 multiplier.
Therefore, no matter _reading_get_ or _reading_update_, it must return agent a value multiplied with platform's scale.

> Bits[15:11] Exponent: The power-of-10 multiplier in two’s-complement format that is applied to the sensor unit specified by the SensorType field.

For example.
In backend, rustvmm reads "**in_anglvel_x_raw**" and return "53", and reads "**in_anglvel_scale**" and return "0.000152716"
then function "**_custom_exponent_**" will returns "-4" here. The real physical data should be "53*0.000152716=0.008093948"

And frontend reads "**in_anglvel_x_scale**" will return "0.000100000" (because custom_exponent is -4). therefore the raw data from "**in_anglvel_x_raw**" should return "80" (So that 80*0.000100000 is closer to the real physical value)

Am I analysis it correct? I will be very honored to refer your opinion.

If yes, I will refactor "_reading_update_" to do same things with "_reading_get_"

Reply:
Yes, I think we understand each other now in this matter.
As for sysfs scale files, does it matter the order in which the files are looked up or is only one of the in_anglevel_y_scale, in_anglevel_scale, and scale present?

Reply:
I am not sure whether order is important, but in theory, it will not exist more than one "scale" sysfs for one axis.
I will research more in kernel, maybe answer can be found out.